src/target/riscv/riscv.c

```diff
@@ -1042,7 +1042,9 @@ static int add_trigger(struct target *target, struct trigger *trigger)
 			break;
 	} while (0);
 
-	riscv_set_register(target, GDB_REGNO_TSELECT, tselect);
+	if (riscv_set_register(target, GDB_REGNO_TSELECT, tselect) != ERROR_OK &&
+			ret == ERROR_OK)
+		return ERROR_FAIL;
```

Comment:
While making changes here: It may not be clear to the reader why TSELECT must be restored here. I would recommend adding a short comment if you like.